### ISSUE TITLE
parameters.grad to be None under certain conditions

### ISSUE THREAD

**Issue by seanywang0408** [pytorch/pytorch#35672]
## To Reproduce

The gradients of all parameters are None after calling backward function, when the following three conditions are fulfilled:
1. model moved to multi-gpus by DataParallel
2. a function to warp the model.forward()
3. two variables are returned by model.forward()

The following code can reproduce the error

```
import torch
import torch.nn as nn

class TestModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(10,2)
        self.fc2 = nn.Linear(10,2)

    def forward(self, x):
        return self.fc1(x), self.fc2(x)

def model_forward(model, x):
    o1, o2 = model(x) # two variables are returned
    return o1

x = torch.rand(2, 10)
model = TestModel()
model = torch.nn.DataParallel(model).cuda() # to multi-gpu

x = x.cuda()
o1 = model_forward(model, x) # a function to warp the forward process
o1.sum().backward()

for p in model.parameters():
    print(p.grad)

```

## Expected behavior

get gradients of model.fc1.weights and model.fc1.bias

## Environment
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-5)
CMake version: version 3.16.2

Python version: 3.7
Is CUDA available: Yes
GPU models and configuration:  two Tesla P40s

Nvidia driver version: 396.44

## Additional context

I tried to remove each of the three conditions and see what would happen. I post the observations below:
1. not to move the model to multi-gpu: fc1 has gradients while the gradients of fc2 is 0
2. not to warp model.forward() with a function: fc1 has gradients while the gradients of fc2 is None
3. only return output of fc1 in model.forward(): fc1 has gradients while the gradients of fc2 is None

I'm curious why the behaviour of the 1st try is different with the rest two.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar

**Comment by mrshenli** [comment]
@seanywang0408  I tried this with the latest master, looks like it has been fixed. 

```
tensor([[0.4842, 1.5574, 1.1879, 0.5454, 0.9086, 1.2837, 1.3273, 0.4979, 0.3307,
         0.4577],
        [0.4842, 1.5574, 1.1879, 0.5454, 0.9086, 1.2837, 1.3273, 0.4979, 0.3307,
         0.4577]], device='cuda:0')
tensor([2., 2.], device='cuda:0')
tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]], device='cuda:0')
tensor([0., 0.], device='cuda:0')
```

I am closing this issue. If you still see the same error on master. Please feel free to reopen.
